File: mfdview/src-tauri/src/raydb.rs
//! Protocole RayDB (TCP 23333) — portage Rust de `raydb_decode.py`.
//!
//! Le module ne connaît que le protocole : construire HELLO et SUBSCRIBE,
//! découper le flux en trames, décoder un UPDATE. Le rangement des valeurs et
//! leur envoi vers la page sont dans `main.rs`.
//!
//! Trame :  [u32 len][u32 msg_type][u8 op][u32 path_len][u32 pad][path][valeur]
//! `len` couvre tout ce qui suit le champ len lui-même. Tout est little-endian.
//!
//! Bloc valeur : [3 octets réservés][u32 type][valeur][4 octets de queue].
//!   0 bool(1)  1 i8(1)  2 i16(2)  3 i32(4)  4 i64(8)  7 u32(4)
//!   9 f32(4)  10 f64(8)  11 chaîne [u64 len][octets]
//!   13 liste : [u64 n] puis n × ([u32 type][valeur])
//!   14 table : [u64 n] puis n × ([u64 klen][clé][u32 type][valeur])
//! Une valeur suit immédiatement son type ; seules les chaînes portent une
//! longueur. Voir « docs/2. protocole-raydb-23333.md » §5.1.

use serde_json::{json, Map, Value};

pub const RAYDB_PORT: u16 = 23333;

const MSG_TYPE: u32 = 1;
const OP_SUBSCRIBE: u8 = 3;
const OP_UPDATE: u8 = 4;
const OP_HELLO: u8 = 7;
const TYPE_STR: u32 = 11; // [u64 len][octets]
const TYPE_LIST: u32 = 13; // [u64 n] puis n valeurs typées
const TYPE_NAMED: u32 = 14; // table [u64 n] puis n × [clé, valeur]

fn u32le(b: &[u8], o: usize) -> Option<u32> {
    Some(u32::from_le_bytes(b.get(o..o + 4)?.try_into().ok()?))
}

fn u64le(b: &[u8], o: usize) -> Option<u64> {
    Some(u64::from_le_bytes(b.get(o..o + 8)?.try_into().ok()?))
}

/// Le MFD laisse traîner de vieux octets derrière le NUL final : on coupe.
fn cstr(b: &[u8]) -> String {
    let end = b.iter().position(|&c| c == 0).unwrap_or(b.len());
    String::from_utf8_lossy(&b[..end]).into_owned()
}

/// Trame de requête. Les 3 octets « réservés » sont opaques : repris tels quels
/// des captures (ils diffèrent selon l'opcode) pour rester acceptés par le MFD.
fn request(op: u8, path: &str, reserved: [u8; 3]) -> Vec<u8> {
    let p = path.as_bytes();
    let mut body = Vec::with_capacity(29 + p.len());
    body.extend_from_slice(&MSG_TYPE.to_le_bytes());
    body.push(op);
    body.extend_from_slice(&(p.len() as u32).to_le_bytes());
    body.extend_from_slice(&0u32.to_le_bytes()); // pad / flags
    body.extend_from_slice(p);
    body.extend_from_slice(&reserved); // bloc valeur « vide »
    body.extend_from_slice(&7u32.to_le_bytes());
    body.extend_from_slice(&0u64.to_le_bytes());

    let mut frame = Vec::with_capacity(4 + body.len());
    frame.extend_from_slice(&(body.len() as u32).to_le_bytes());
    frame.extend_from_slice(&body);
    frame
}

/// Le client s'annonce — mêmes octets que le vrai client Raymarine.
pub fn hello() -> Vec<u8> {
    request(OP_HELLO, "RayDBRemoteClient", [0x00, 0x01, 0x01])
}

/// Abonnement à un chemin (`data/#` = tout le sous-arbre navigation).
pub fn subscribe(path: &str) -> Vec<u8> {
    request(OP_SUBSCRIBE, path, [0x01, 0x00, 0x00])
}

/// Extrait du tampon toutes les trames complètes, et ne laisse que le reste.
/// Une lecture TCP peut couper au milieu d'une trame ou en livrer plusieurs.
pub fn take_frames(buf: &mut Vec<u8>) -> Vec<Vec<u8>> {
    let mut frames = Vec::new();
    let mut o = 0;
    while buf.len() - o >= 4 {
        let len = match u32le(buf, o) {
            Some(v) => v as usize,
            None => break,
        };
        if len < 13 {
            break; // longueur invalide : flux désynchronisé, on n'avance plus
        }
        let total = 4 + len;
        if buf.len() - o < total {
            break; // trame incomplète : on attend la suite
        }
        frames.push(buf[o..o + total].to_vec());
        o += total;
    }
    buf.drain(..o);
    frames
}

/// Lit la valeur de type `vtype` à l'offset `o`, et rend l'offset de fin.
///
/// L'énumération des types se lit comme une suite d'entiers de largeur
/// croissante, puis les composites (§5.1 de la spec). Les listes et les tables
/// s'imbriquent, d'où la récursion. 5, 6, 8 et 12 n'ont jamais été observés ;
/// leur place dans la suite est déduite de leurs voisins.
fn read_value(b: &[u8], o: usize, vtype: u32) -> Option<(Value, usize)> {
    let fixed = |n: usize| -> Option<&[u8]> { b.get(o..o + n) };
    match vtype {
        0 => Some((json!(*b.get(o)? != 0), o + 1)),
        1 => Some((json!(*b.get(o)? as i8), o + 1)),
        2 => Some((json!(i16::from_le_bytes(fixed(2)?.try_into().ok()?)), o + 2)),
        3 => Some((json!(i32::from_le_bytes(fixed(4)?.try_into().ok()?)), o + 4)),
        4 => Some((json!(i64::from_le_bytes(fixed(8)?.try_into().ok()?)), o + 8)),
        5 => Some((json!(*b.get(o)?), o + 1)),
        6 => Some((json!(u16::from_le_bytes(fixed(2)?.try_into().ok()?)), o + 2)),
        7 => Some((json!(u32le(b, o)?), o + 4)),
        8 => Some((json!(u64le(b, o)?), o + 8)),
        9 => Some((
            json!(f32::from_le_bytes(fixed(4)?.try_into().ok()?) as f64),
            o + 4,
        )),
        10 => Some((json!(f64::from_le_bytes(fixed(8)?.try_into().ok()?)), o + 8)),
        TYPE_STR => {
            let n = u64le(b, o)? as usize;
            Some((json!(cstr(b.get(o + 8..o + 8 + n)?)), o + 8 + n))
        }
        TYPE_LIST | TYPE_NAMED => {
            let n = u64le(b, o)? as usize;
            if n > b.len() {
                return None; // compteur aberrant : bloc tronqué ou mal cadré
            }
            let mut o = o + 8;
            let mut items = Vec::new();
            let mut map = Map::new();
            for _ in 0..n {
                let mut key = String::new();
                if vtype == TYPE_NAMED {
                    let klen = u64le(b, o)? as usize;
                    o += 8;
                    key = cstr(b.get(o..o + klen)?);
                    o += klen;
                }
                let (value, end) = read_value(b, o + 4, u32le(b, o)?)?;
                o = end;
                if vtype == TYPE_NAMED {
                    map.insert(key, value);
                } else {
                    items.push(value);
                }
            }
            let out = if vtype == TYPE_NAMED {
                Value::Object(map)
            } else {
                Value::Array(items)
            };
            Some((out, o))
        }
        _ => None,
    }
}

/// Bloc valeur d'un UPDATE : [3 réservés][u32 type][valeur][4 octets de queue].
fn typed(b: &[u8]) -> Option<Value> {
    if b.len() < 8 || b[0..3] != [0, 0, 0] {
        return None;
    }
    let vtype = u32le(b, 3)?;
    // Table à entrée unique — la forme des `diag/…` : le nom réencodé y double
    // le plus souvent le dernier segment du chemin, que la page a déjà comme
    // clé, donc on ne rend que la valeur. Au-delà d'une entrée, la table est
    // rendue telle quelle : n'en garder que la première perdait le reste en
    // silence (`data/exportedCamera/…` porte six champs).
    if vtype == TYPE_NAMED && u64le(b, 7)? == 1 {
        let klen = u64le(b, 15)? as usize;
        let o = 23 + klen;
        return Some(read_value(b, o + 4, u32le(b, o)?)?.0);
    }
    Some(read_value(b, 7, vtype)?.0)
}

/// (chemin, valeur) d'une trame UPDATE ; None pour les autres opcodes.
pub fn decode_update(frame: &[u8]) -> Option<(String, Value)> {
    if frame.len() < 17 || frame[8] != OP_UPDATE {
        return None;
    }
    let plen = u32le(frame, 9)? as usize;
    let path = cstr(frame.get(17..17 + plen)?);
    Some((path, typed(frame.get(17 + plen..)?)?))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Blocs valeur relevés tels quels dans `pcap/boat-c_*.pcap` — un par type.
    fn block(hex: &str) -> Vec<u8> {
        (0..hex.len())
            .step_by(2)
            .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).unwrap())
            .collect()
    }

    #[test]
    fn scalaires() {
        // Diagnostics/CrashLogCount_3712024258 = 337 (i32, type 3)
        assert_eq!(typed(&block("000000030000005101000000000000")), Some(json!(337)));
        // Settings/Data/-/10/2/-/-/-/- = 1 (i64, type 4)
        assert_eq!(
            typed(&block("00000004000000010000000000000000000000")),
            Some(json!(1))
        );
    }

    #[test]
    fn liste() {
        // AvailableCartoVendorClient/availableSysCarto = [1]
        let b = block("0000000d000000010000000000000007000000010000000000000000");
        assert_eq!(typed(&b), Some(json!([1])));
    }

    #[test]
    fn table_a_plusieurs_entrees() {
        // Settings/Data/-/3/89/-/-/-/- relevé dans pcap/boat-c_axiom7.pcap :
        // cinq entrées i32, que l'ancien décodeur réduisait à la première.
        let b = block(concat!(
            "0000000e00000005000000000000000300000000000000484c",
            "5303000000000000000300000000000000484d530300000000",
            "00000003000000000000004c4c530300000000000000030000",
            "00000000004c4d530300000000000000060000000000000056",
            "656e646f72030000000100000000000000"
        ));
        assert_eq!(
            typed(&b),
            Some(json!({"HLS": 0, "HMS": 0, "LLS": 0, "LMS": 0, "Vendor": 1}))
        );
    }

    #[test]
    fn table_a_entree_unique_rend_la_valeur_nue() {
        // diag/… : [count=1][klen=3]["HLS"][type 3][valeur 0]
        let b = block("0000000e00000001000000000000000300000000000000484c530300000000000000");
        assert_eq!(typed(&b), Some(json!(0)));
    }
}

File: raynmea/internal/gateway/raydb.go
// raydb.go — protocole RayDB (TCP 23333) : construction des requêtes, lecture
// des trames, décodage du bloc valeur.
//
// Portage de `raydb_decode.py` / `raydb_client.py`, spec dans
// « docs/2. protocole-raydb-23333.md ». Une trame vaut, en little-endian :
//
//	[u32 len][u32 msg_type=1][u8 op][u32 path_len][u32 pad][path][bloc valeur]
//
// où `len` couvre tout ce qui suit le champ len lui-même. Le bloc valeur est
// [3 octets réservés][u32 type][valeur][4 octets de queue], avec :
//
//	0 bool(1)  1 i8(1)  2 i16(2)  3 i32(4)  4 i64(8)  7 u32(4)
//	9 f32(4)  10 f64(8)  11 chaîne [u64 len][octets]
//	13 liste : [u64 n] puis n × ([u32 type][valeur])
//	14 table : [u64 n] puis n × ([u64 klen][clé][u32 type][valeur])
//
// Une valeur suit immédiatement son type ; seules les chaînes portent une
// longueur. Voir § 5.1 de la spec.
package gateway

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"strconv"
	"strings"
)

const (
	raydbPort = 23333

	msgType     = 1
	opSubscribe = 3
	opUpdate    = 4
	opHello     = 7

	// Nom sous lequel le client s'annonce — celui du vrai client Raymarine.
	// Le MFD n'en fait rien, mais on reste à l'octet près sur les captures.
	clientName = "RayDBRemoteClient"

	// Un en-tête complet fait 17 octets ; aucune trame légitime n'approche 1 Mo,
	// d'où le garde-fou (une longueur aberrante veut dire flux désynchronisé).
	frameHeaderLen = 17
	maxFrameLen    = 1 << 20
)

// Types du bloc valeur (§5.1 de la spec). L'énumération se lit comme une suite
// d'entiers de largeur croissante, puis les composites. 0x05, 0x06, 0x08 et
// 0x0C n'ont jamais été observés ; leur place est déduite de leurs voisins.
const (
	typeBool  = 0x00
	typeI8    = 0x01
	typeI16   = 0x02
	typeI32   = 0x03
	typeI64   = 0x04
	typeU8    = 0x05
	typeU16   = 0x06
	typeU32   = 0x07
	typeU64   = 0x08
	typeF32   = 0x09
	typeF64   = 0x0A
	typeStr   = 0x0B
	typeList  = 0x0D // [u64 n] puis n × ([u32 type][valeur])
	typeNamed = 0x0E // table : [u64 n] puis n × ([u64 klen][clé][u32 type][valeur])
)

// ------------------------------------------------------ requêtes montantes ---

// buildFrame assemble une trame RayDB complète.
func buildFrame(op byte, path string, tail []byte) []byte {
	p := []byte(path)
	body := make([]byte, 0, 13+len(p)+len(tail))
	body = binary.LittleEndian.AppendUint32(body, msgType)
	body = append(body, op)
	body = binary.LittleEndian.AppendUint32(body, uint32(len(p)))
	body = binary.LittleEndian.AppendUint32(body, 0) // pad / flags
	body = append(body, p...)
	body = append(body, tail...)

	out := binary.LittleEndian.AppendUint32(make([]byte, 0, 4+len(body)), uint32(len(body)))
	return append(out, body...)
}

// requestTail rend le bloc valeur « vide » des requêtes :
// [3 réservés][u32 type=0x07][u64 0]. Les 3 octets réservés sont opaques —
// repris tels quels des captures, où ils diffèrent selon l'opcode.
func requestTail(reserved ...byte) []byte {
	tail := append([]byte{}, reserved...)
	tail = binary.LittleEndian.AppendUint32(tail, typeU32)
	return binary.LittleEndian.AppendUint64(tail, 0)
}

// buildHello : le client s'annonce (mêmes octets que le client Raymarine).
func buildHello(name string) []byte {
	return buildFrame(opHello, name, requestTail(0x00, 0x01, 0x01))
}

// buildSubscribe : abonnement à un chemin ou à un sous-arbre (« data/# »).
func buildSubscribe(path string) []byte {
	return buildFrame(opSubscribe, path, requestTail(0x01, 0x00, 0x00))
}

// ------------------------------------------------------------ lecture flux ---

// readFrame lit une trame et rend (op, chemin, bloc valeur). Le découpage se
// fait sur le champ de longueur : une trame peut être scindée entre plusieurs
// segments TCP, et plusieurs trames tenir dans un seul.
func readFrame(r *bufio.Reader) (op byte, path string, block []byte, err error) {
	var head [4]byte
	if _, err = io.ReadFull(r, head[:]); err != nil {
		return 0, "", nil, err
	}
	n := binary.LittleEndian.Uint32(head[:])
	if n < frameHeaderLen-4 || n > maxFrameLen {
		return 0, "", nil, fmt.Errorf("longueur de trame aberrante : %d", n)
	}
	body := make([]byte, n)
	if _, err = io.ReadFull(r, body); err != nil {
		return 0, "", nil, err
	}
	op = body[4]
	plen := int(binary.LittleEndian.Uint32(body[5:]))
	if plen < 0 || 13+plen > len(body) {
		return 0, "", nil, fmt.Errorf("chemin de %d octets dans une trame de %d", plen, n)
	}
	return op, latin1(body[13 : 13+plen]), body[13+plen:], nil
}

// --------------------------------------------------------- bloc valeur -------

type valueKind uint8

const (
	kindNone valueKind = iota
	kindBool
	kindInt // entiers autres que u32 : i8, i16, i32, i64, u8, u16, u64
	kindU32
	kindF32
	kindF64
	kindStr
	kindList
	kindMap
)

// Field est une entrée nommée d'une table (type 0x0e). Les entrées sont gardées
// en tranche et non en map : le flux les donne dans un ordre, que l'affichage
// doit rendre stable.
type Field struct {
	Name  string
	Value Value
}

// Value est une valeur RayDB décodée, dans son type d'origine. Le float32 est
// gardé sur 32 bits : converti en double, il s'étale en chiffres que le MFD n'a
// jamais envoyés (0.03009999915957451 pour un 0,0301 émis).
type Value struct {
	Kind   valueKind
	Name   string // nom réencodé d'une table à entrée unique, s'il apporte quelque chose
	Bool   bool
	Int    int64
	U32    uint32
	F32    float32
	F64    float64
	Str    string
	Items  []Value // kindList
	Fields []Field // kindMap
}

// Number rend la valeur numérique de v, et si elle en a une. Le MFD publie NaN
// pour « pas de mesure » (data/cog/stable en donne) : ce n'est pas une valeur,
// ok vaut alors false — comme le `None` du client Python.
func (v Value) Number() (float64, bool) {
	var f float64
	switch v.Kind {
	case kindInt:
		return float64(v.Int), true
	case kindU32:
		return float64(v.U32), true
	case kindF32:
		f = float64(v.F32)
	case kindF64:
		f = v.F64
	default:
		return 0, false
	}
	return f, !math.IsNaN(f) && !math.IsInf(f, 0)
}

// String rend la valeur telle qu'on l'affiche. Les flottants sortent dans leur
// écriture décimale la plus courte qui les redonne bit pour bit — à la largeur
// où ils ont été reçus, 32 ou 64 bits.
func (v Value) String() string {
	var s string
	switch v.Kind {
	case kindBool:
		s = "false"
		if v.Bool {
			s = "true"
		}
	case kindInt:
		s = strconv.FormatInt(v.Int, 10)
	case kindU32:
		s = strconv.FormatUint(uint64(v.U32), 10)
	case kindF32:
		s = strconv.FormatFloat(float64(v.F32), 'g', -1, 32)
	case kindF64:
		s = strconv.FormatFloat(v.F64, 'g', -1, 64)
	case kindStr:
		s = v.Str
	case kindList:
		parts := make([]string, len(v.Items))
		for i, item := range v.Items {
			parts[i] = item.String()
		}
		s = "[" + strings.Join(parts, ", ") + "]"
	case kindMap:
		parts := make([]string, len(v.Fields))
		for i, fld := range v.Fields {
			parts[i] = fld.Name + ": " + fld.Value.String()
		}
		s = "{" + strings.Join(parts, ", ") + "}"
	default:
		return ""
	}
	if v.Name != "" {
		return v.Name + " = " + s
	}
	return s
}

// decodeValue décode le bloc valeur d'un UPDATE. `leaf` est le dernier segment
// du chemin : une table à entrée unique (type 0x0e, la forme des `diag/…`)
// réencode le nom du champ, qui double le plus souvent ce segment — on ne garde
// le nom que s'il en diffère, auquel cas il porte une information que le chemin
// n'a pas. Au-delà d'une entrée, la table est rendue telle quelle : n'en garder
// que la première perdait les autres en silence.
func decodeValue(b []byte, leaf string) (Value, bool) {
	if len(b) < 8 || b[0] != 0 || b[1] != 0 || b[2] != 0 {
		return Value{}, false
	}
	vtype := binary.LittleEndian.Uint32(b[3:])
	if vtype == typeNamed && len(b) >= 15 && binary.LittleEndian.Uint64(b[7:]) == 1 {
		// [u64 count=1 @7][u64 klen @15][nom @23][u32 type][valeur]
		if len(b) < 23 {
			return Value{}, false
		}
		klen := binary.LittleEndian.Uint64(b[15:])
		if klen > uint64(len(b)-23) {
			return Value{}, false
		}
		o := 23 + int(klen)
		if len(b) < o+4 {
			return Value{}, false
		}
		v, _, ok := decodeAt(b, o+4, binary.LittleEndian.Uint32(b[o:]))
		if !ok {
			return Value{}, false
		}
		if name := latin1(b[23:o]); name != leaf {
			v.Name = name
		}
		return v, true
	}
	v, _, ok := decodeAt(b, 7, vtype)
	return v, ok
}

// Largeur des types à taille fixe, en octets.
var scalarWidth = map[uint32]int{
	typeBool: 1, typeI8: 1, typeU8: 1, typeI16: 2, typeU16: 2,
	typeI32: 4, typeU32: 4, typeF32: 4, typeI64: 8, typeU64: 8, typeF64: 8,
}

// decodeAt lit la valeur de type `vtype` à l'offset `o` et rend l'offset de fin.
// La valeur suit immédiatement son type : seules les chaînes portent une
// longueur. Listes et tables s'imbriquent, d'où la récursion.
func decodeAt(b []byte, o int, vtype uint32) (Value, int, bool) {
	if w, fixed := scalarWidth[vtype]; fixed {
		if o < 0 || len(b) < o+w {
			return Value{}, 0, false
		}
		end := o + w
		switch vtype {
		case typeBool:
			return Value{Kind: kindBool, Bool: b[o] != 0}, end, true
		case typeI8:
			return Value{Kind: kindInt, Int: int64(int8(b[o]))}, end, true
		case typeU8:
			return Value{Kind: kindInt, Int: int64(b[o])}, end, true
		case typeI16:
			return Value{Kind: kindInt, Int: int64(int16(binary.LittleEndian.Uint16(b[o:])))}, end, true
		case typeU16:
			return Value{Kind: kindInt, Int: int64(binary.LittleEndian.Uint16(b[o:]))}, end, true
		case typeI32:
			return Value{Kind: kindInt, Int: int64(int32(binary.LittleEndian.Uint32(b[o:])))}, end, true
		case typeU32:
			return Value{Kind: kindU32, U32: binary.LittleEndian.Uint32(b[o:])}, end, true
		case typeI64, typeU64:
			return Value{Kind: kindInt, Int: int64(binary.LittleEndian.Uint64(b[o:]))}, end, true
		case typeF32:
			return Value{Kind: kindF32, F32: math.Float32frombits(binary.LittleEndian.Uint32(b[o:]))}, end, true
		case typeF64:
			return Value{Kind: kindF64, F64: math.Float64frombits(binary.LittleEndian.Uint64(b[o:]))}, end, true
		}
	}

	switch vtype {
	case typeStr:
		if o < 0 || len(b) < o+8 {
			return Value{}, 0, false
		}
		n := binary.LittleEndian.Uint64(b[o:])
		if n > uint64(len(b)-o-8) {
			return Value{}, 0, false
		}
		return Value{Kind: kindStr, Str: latin1(b[o+8 : o+8+int(n)])}, o + 8 + int(n), true

	case typeList, typeNamed:
		if o < 0 || len(b) < o+8 {
			return Value{}, 0, false
		}
		n := binary.LittleEndian.Uint64(b[o:])
		if n > uint64(len(b)) { // compteur aberrant : bloc tronqué ou mal cadré
			return Value{}, 0, false
		}
		o += 8
		v := Value{Kind: kindList}
		if vtype == typeNamed {
			v = Value{Kind: kindMap}
		}
		for i := uint64(0); i < n; i++ {
			name := ""
			if vtype == typeNamed {
				if len(b) < o+8 {
					return Value{}, 0, false
				}
				klen := binary.LittleEndian.Uint64(b[o:])
				o += 8
				if klen > uint64(len(b)-o) {
					return Value{}, 0, false
				}
				name = latin1(b[o : o+int(klen)])
				o += int(klen)
			}
			if len(b) < o+4 {
				return Value{}, 0, false
			}
			inner, end, ok := decodeAt(b, o+4, binary.LittleEndian.Uint32(b[o:]))
			if !ok {
				return Value{}, 0, false
			}
			o = end
			if vtype == typeNamed {
				v.Fields = append(v.Fields, Field{Name: name, Value: inner})
			} else {
				v.Items = append(v.Items, inner)
			}
		}
		return v, o, true
	}
	return Value{}, 0, false
}

// latin1 rend la chaîne portée par ces octets, coupée au premier NUL — au-delà,
// c'est un tampon réutilisé non nettoyé. Le protocole est en latin1 : chaque
// octet est un point de code, là où Go supposerait de l'UTF-8.
func latin1(b []byte) string {
	if i := bytes.IndexByte(b, 0); i >= 0 {
		b = b[:i]
	}
	for _, c := range b {
		if c > 0x7f {
			r := make([]rune, len(b))
			for i, c := range b {
				r[i] = rune(c)
			}
			return string(r)
		}
	}
	return string(b)
}
